src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs

```diff
@@ -2315,6 +2330,51 @@ internal bool LoadModuleManifestData(ExternalScriptInfo scriptInfo, ManifestProc
                 if (bailOnFirstError) return null;
             }
 
+            // On Windows, we want to include any modules under %WINDIR%\System32\WindowsPowerShell\v1.0\Modules
+            // that have declared compatibility with PS Core (or if the check is skipped)
+            IEnumerable<string> inferredCompatiblePSEditions = compatiblePSEditions ?? DefaultCompatiblePSEditions;
+            if (!IsPSEditionCompatible(moduleManifestPath, inferredCompatiblePSEditions, out bool isOnSystem32ModulePath))
+            {
+                containedErrors = true;
+
+                if (writingErrors)
+                {
+                    message = StringUtil.Format(
+                        Modules.PSEditionNotSupported,
+                        moduleManifestPath,
+                        PSVersionInfo.PSEdition,
+                        String.Join(',', inferredCompatiblePSEditions));
+
+                    InvalidOperationException ioe = new InvalidOperationException(message);
+
+                    ErrorRecord er = new ErrorRecord(
+                        ioe,
+                        nameof(Modules) + "_" + nameof(Modules.PSEditionNotSupported),
+                        ErrorCategory.ResourceUnavailable,
+                        moduleManifestPath);
+
+                    WriteError(er);
+                }
+
+                if (bailOnFirstError)
+                {
+                    // If we're trying to load the module, return null so that caches
+                    // are not polluted
+                    if (manifestProcessingFlags.HasFlag(ManifestProcessingFlags.LoadElements))
+                    {
+                        return null;
+                    }
+
+                    // If we return null with Get-Module, a fake module info will be created. Since
+                    // we want to suppress output of the module, we need to do that here.
+                    return new PSModuleInfo(moduleManifestPath, context: null, sessionState: null)
```

Comment:
Why do we need to return a dummy module info object in case of `Get-Module`? If it's not compatible, `Get-Module` will not return it anyway, seems no need to return a dummy module info.

Reply:
The calling method in the context of `Get-Module` will take a `null` result and decide to make a dummy `PSModuleInfo` object that doesn't say "I'm incompatible", so `Get-Module` will show it when it fails the check.

So this creates the fake `PSModuleInfo` earlier with the `IsLoadedFromCompatibilityCheckedPath` set.

But it's not a great piece of code, just the cheapest fix at the moment.

Reply:
Discussed with @rjmholt offline on this. The current behavior for `Get-Module -ListAvailable` is that if we have a problem parsing the manifest, we just create a dummy `PSModuleInfo` using the file path, and still display that module to the user.

This behavior doesn't work with the new behavior where we want to hide modules that are incompatible -- say we got parsing error for a module manifest under system 32, should we return a dummy module for that file or not? We got a parsing error, so we don't know if that module has `Core` declared or not.

A more suitable behavior would be don't return a dummy module in case we got a parsing error, because practically, that module is not available since it cannot be imported. We defer solving this along with the other two issues described in my comments.

Reply:
Please open tracking and reference issues.

Reply:
Please see https://github.com/PowerShell/PowerShell/issues/7298